experimental/dds/tree2/src/feature-libraries/editable-tree/editableTreeBinder.ts

```diff
@@ -0,0 +1,930 @@
+
+/**
+ * A step in a bind path
+ *
+ * @alpha
+ */
+export interface PathStep {
+	readonly field: FieldKey;
+	readonly index?: number;
+}
+/**
+ * A node in a bind path
+ *
+ * @alpha
+ */
+export interface BindTree extends PathStep {
+	readonly children: Map<FieldKey, BindTree>;
+}
+
+/**
+ * Index symbol for syntax tree
+ *
+ * @alpha
+ */
+export const indexSymbol = Symbol("editable-tree-binder:index");
+
+/**
+ * A syntax node for the bind language
+ *
+ * @alpha
+ */
+export interface BindSyntaxTree {
+	readonly [indexSymbol]?: number;
+	readonly [key: string]: boolean | BindSyntaxTree | number | undefined;
+}
+
+/**
+ * A down path
+ *
+ * @alpha
+ */
+export type DownPath = PathStep[];
+
+/**
+ * A bind path
+ *
+ * @alpha
+ */
+export type BindPath = DownPath;
+
+/**
+ * @alpha
+ */
+export type BindingContext = DeleteBindingContext | InsertBindingContext | SetValueBindingContext;
+
+/**
+ * @alpha
+ */
+export type BindingContextQueue = BindingContext[];
+
+/**
+ * @alpha
+ */
+export const BindingType = {
+	Delete: "delete",
+	Insert: "insert",
+	SetValue: "setValue",
+	Invalidation: "invalidation",
+	Batch: "batch",
+} as const;
+
+/**
+ * @alpha
+ */
+export type BindingContextType = typeof BindingType[keyof typeof BindingType];
+
+/**
+ * @alpha
+ */
+export interface AbstractBindingContext {
+	readonly type: BindingContextType;
+}
+
+/**
+ * @alpha
+ */
+export interface DeleteBindingContext extends AbstractBindingContext {
+	readonly type: typeof BindingType.Delete;
+	readonly path: UpPath;
+	readonly count: number;
+}
+
+/**
+ * @alpha
+ */
+export interface InsertBindingContext extends AbstractBindingContext {
+	readonly type: typeof BindingType.Insert;
+	readonly path: UpPath;
+	readonly content: ProtoNodes;
+}
+
+/**
+ * @alpha
+ */
+export interface SetValueBindingContext extends AbstractBindingContext {
+	readonly type: typeof BindingType.SetValue;
+	readonly path: UpPath;
+	readonly value: TreeValue;
+}
+
+/**
+ * @alpha
+ */
+export interface InvalidationBindingContext extends AbstractBindingContext {
+	readonly type: typeof BindingType.Invalidation;
+}
+
+/**
+ * @alpha
+ */
+export interface BatchBindingContext extends AbstractBindingContext {
+	readonly type: typeof BindingType.Batch;
+	readonly events: BindingContext[];
+}
+
+abstract class AbstractPathVisitor implements PathVisitor {
+	protected readonly registeredPaths: Map<
+		BindingContextType,
+		Map<BindPath, Set<(...args: unknown[]) => any>>
+	> = new Map();
+	public constructor(protected readonly options: BinderOptions) {}
+	public abstract onDelete(path: UpPath, count: number): void;
+	public abstract onInsert(path: UpPath, content: ProtoNodes): void;
+	public abstract onSetValue(path: UpPath, value: TreeValue): void;
+
+	protected matchesPath(visitPath: BindPath | undefined, downPath: BindPath): boolean {
+		if (visitPath !== undefined) {
+			const matchPolicy = this.options.matchPolicy;
+			switch (matchPolicy) {
+				case "subtree":
```

Comment:
Probably a preliminary optimization, but doing this string comparison for every hop the visitor does seems very expensive. Maybe we could use some other mechanism, like a callback or inheritance for this?

Reply:
Removed by refactoring